### PR TITLE
Break ``intersphinx`` into a package

### PR DESCRIPTION
**Pull request by chrisjsewell**
This change completely maintains the public API and includes no changes in code logic

Internally though it split the code into three, independent, concerns:

1. Loading inventory data into an internal representation
2. Using the internally represented inventories to resolve references
3. Providing command-line utilities

This makes the logic clearer and more maintainable for future changes

It also remove this module from the `ruff format` exclude list

### CODE REVIEW

File: sphinx/ext/intersphinx/_load.py
Comment (picnixz) [inline sphinx/ext/intersphinx/_load.py]: Wouldn't the logger name be different then?

**Comment by picnixz** [comment]
> This change completely maintains the public API and includes no changes in code logic

I assume that all public functions were re-exported right?

**Comment by chrisjsewell** [comment]
> I assume that all public functions were re-exported right?

yes, they are all part of `__all__`

File: sphinx/ext/intersphinx/_load.py
Comment (chrisjsewell) [reply sphinx/ext/intersphinx/_load.py]: Ah, I can change the logger name if you want 😅?
although, practically speaking I think it would an extremely niche use case for something to be depending on these logger names (sphinx already suffixes them in `getLogger`)

File: sphinx/ext/intersphinx/__init__.py
Comment (picnixz) [inline sphinx/ext/intersphinx/__init__.py]: Do we put `__all__` before or after imports ? I personally put it at the very top so that when you open a file, you directly see what is being exported but I'm not sure what convention we should take.

Comment (jayaddison) [reply sphinx/ext/intersphinx/__init__.py]: I'd like to see us follow [PEP8 guidance](https://peps.python.org/pep-0008/#module-level-dunder-names) on this.

However: doing that would be best achieved with linting support.  I've filed a feature request #12179 for that.. but it may require retaining `flake8`.

Comment (picnixz) [reply sphinx/ext/intersphinx/__init__.py]: I recommend using PEP 8.

File: sphinx/ext/intersphinx/_load.py
Comment (picnixz) [reply sphinx/ext/intersphinx/_load.py]: Yes but we never know. I'm not entirely sure that [my extension](https://github.com/picnixz/sphinx-zeta-suppress) would work if you change it (well it would work but you'll need to change the configuration).

So I think it's better if you use the same logger for the whole intersphinx module (or we should make it backward compatible).

Comment (chrisjsewell) [reply sphinx/ext/intersphinx/_load.py]: fixed

File: sphinx/ext/intersphinx/__init__.py
Comment (chrisjsewell) [reply sphinx/ext/intersphinx/__init__.py]: moved to before imports

File: sphinx/ext/intersphinx/_shared.py
Comment (picnixz) [inline sphinx/ext/intersphinx/_shared.py]: (I'd suggest having a `Final` annotation for this one)

Comment (chrisjsewell) [reply sphinx/ext/intersphinx/_shared.py]: done

**Review by jakobandersen — COMMENTED** [review]
Assuming this basically just moves code around (haven't checked), then I don't see any issues.

**Comment by picnixz** [comment]
I haven't checked again but if nothing changed since I reviewed and if everything that was public is still exported at the same location, I'm fine with that.

Do you want to make another PR where some objects are deprecated and marked as private in 9.x?

**Review by picnixz — APPROVED** [review]

**Comment by chrisjsewell** [comment]
@AA-Turner please don't merge this as is yet 😅 , 
I haven't integrated https://github.com/sphinx-doc/sphinx/pull/12193 into the new code

**Comment by AA-Turner** [comment]
> I haven't integrated #12193 into the new code

I just have; I re-split on your delineations from the current state of intersphinx.py (but kept you as the author / credit via rewriting the commit metadata).

A

**Comment by chrisjsewell** [comment]
> I just have; I re-split on your delineations from the current state of intersphinx.py (but kept you as the author / credit via rewriting the commit metadata

Ok cool cheers, I didn't see that, just wanted to quickly write something before the CLI finished running 😄

but yeh, as long as we have double-checked that the new code matches the old, then all good from my end 👍

**Comment by AA-Turner** [comment]
No worries -- better safe than sorry & all that! I've also broken up the commits that reorganise the package so it's easier to verify that everything has been moved correctly -- and I've deferred any style updates until after this PR, again in an effort to properly verify everything is all well.

A

**Comment by chrisjsewell** [comment]
perfect 👌

**Comment by AA-Turner** [comment]
I intend to rebase-merge this as-is, but if you want me to hold off at all I'm happy to do so @chrisjsewell

**Comment by chrisjsewell** [comment]
> I intend to rebase-merge this as-is, but if you want me to hold off at all I'm happy to do so

No thats great thanks, go ahead!

**Pull request merged**
